New Switchres (CRT Resolution switching) For Retroarch

alphanu1 opened the pull request:
## Description

Fix for current issues and many improvements.

Whole new CRT switching implementation.  This does require external libraries but these are not need for compile or for standard runtime. These are only required when CRT switching is enabled.

SR2 (switchres) is a much better method of resolution switching. It is much cleaner and generates better resolutions than the previous CRTSwitchRes method.

Many old issues and request have been either fixed or implemented.  No longer limited to the primary display, windows dynamic resolutions available. Better and more stable multi monitor support.

Currently, all original functionality except x centering is available. This may be implemented at some point.

Requires libswitchres.so or libswitchres.dll for functionality available from [Switchres](https://github.com/antonioginer/switchres) precompiled windows versions can be made available.

A switchres.ini is not required.  However, without it there are some options that can not be chosen from within ReatroArch UI. 

Thorough testing has been done by the RetroArch, Switchres and GroovyArcade community. With outstanding results 😊

## Related Issues

#12391
#12244

## Reviewers

@hunterk 
@twinaphex

twinaphex commented:
Is this optional and the old version can also still be used? I don't want RetroArch to be entirely dependent on an external library for this kind of functionality. I am fine however if it is an option and the old method can also still be used.

alphanu1 commented:
Currently, no. 

This new switching implimantation is far superior to the previous. Cuttently, the old method is not working in 1.91 and 1.92. The new method is. 

It might be possible to get the old version working again. And then having a toggle to choose between them. However, considering the limitation of the old method and inferior functionality, its not really worth spending time doing!?

Can we consider supplying the libs with the downloads or including the source at compile time?

twinaphex commented:
Can you find a way to put this switchres version in the deps/ dir of RetroArch and then bake it into RetroArch without running it as a dynamically loadable library? We could maybe go about it that way then.

My main concern is just having to source these external libraries, I'd prefer it is all baked-in instead.

alphanu1 commented:
@twinaphex we are looking into some solutions here. I do feel though that adding it as a dep would hinder the continuous development lifecycle for RetroArch/switchers. Any improvement would have to be manually added and merged each time. Unless you are thinking of a different way of doing this? This is obviously not out of the question. However, I do not think its ideal.

Maybe you could consider adding a submodule for switchres. Which would mean each release would always be up to date with any changes or improvements. Or alternatively, keeping switchres out of RA compile time. You could allow the switchres library to be part of the content downloader? This would allow any user to downlead the latest switchres build at any time. Giving them instant access to the very latest improvements and the user would not need to look elsewhere else to find them.

FYI, I just fixed fixed a compile error for some environments and rebased. I still see some C++ comments in the switchres wrapper which are causing issues. I will look into these too.

alphanu1 commented:
Temporarily removed SR2 logging to temporarily fix c90 compile issues

LarsenDX commented:
Regarding the Linux code, please ensure that switchres not only works with libXrandr (X11) but also with DRM/KMS so we can have Switchres on Generic X86_64 (PC) in Lakka-libreELEC!

LarsenDX commented:
You will run into issues in DRM/KMS due to the way retroarch currently controls DRM/KMS.

./switchres 320 240 60 -s
DRM/KMS: <1> (init) [ERROR] limited DRM rights on this screen
Switchres: Modeline "320x240_60 15.660000KHz 60.000000Hz" 6.514560 320 333 364 416 240 242 245 261   -hsync -vsync
**DRM/KMS: <1> (add_mode) [ERROR] cannot attach the mode to the crtc 40 frame buffer 65**
Press ENTER to exit...

substring commented:
> Regarding the Linux code, please ensure that switchres not only works with libXrandr (X11) but also with DRM/KMS so we can have Switchres on Generic X86_64 (PC) in Lakka-libreELEC!

KMS is a MUCH more complicated task for a bunch of reasons. As of now, it's impossible, from switchres, to configure a CRTC (RA "owns" it an configures it itself) and due to KMS internals, it probably never will. I don't know yet how we're gonna do it as it will probably imply some code specific to KMS inside the CRT Switchres code. I have a plan, just didn't have time yet to start coding that part. On the SR part, the KMS code is really hacky, KMS doesn't work like xrandr at all, and is not that flexible to allow a 3rd party lib modify a CRTC (that KMS Master thing ...)

Considering KMS modeswitching never really worked before, there is no regression here.

LarsenDX commented:
> > Regarding the Linux code, please ensure that switchres not only works with libXrandr (X11) but also with DRM/KMS so we can have Switchres on Generic X86_64 (PC) in Lakka-libreELEC!
> 
> KMS is a MUCH more complicated task for a bunch of reasons. As of now, it's impossible, from switchres, to configure a CRTC (RA "owns" it an configures it itself) and due to KMS internals, it probably never will. I don't know yet how we're gonna do it as it will probably imply some code specific to KMS inside the CRT Switchres code. I have a plan, just didn't have time yet to start coding that part. On the SR part, the KMS code is really hacky, KMS doesn't work like xrandr at all, and is not that flexible to allow a 3rd party lib modify a CRTC (that KMS Master thing ...)
> 
> Considering KMS modeswitching never really worked before, there is no regression here.

I looked into this some more. I understand the difficulty. Correct me if I am wrong. Retroarch in DRM/KMS is the DRM-Master. At startup, Retroarch queries for EDID modelines and tries to match video_fullscreen_x/video_fullscreen_y to a modeline. If it cannot find one, Retroarch quits at startup. E.g. 2560x240, unless you EDID-injected that modeline yourself. So making switchres work on Retroarch in DRM/KMS would imply getting into the DRM/KMS code of Retroarch rather than wrapping libswitchres.so? This also means that the user needs to EDID-inject CRT modelines.

substring commented:
> I looked into this some more. I understand the difficulty. Correct me if I am wrong. Retroarch in DRM/KMS is the DRM-Master. At startup, Retroarch queries for EDID modelines and tries to match video_fullscreen_x/video_fullscreen_y to a modeline. If it cannot find one, Retroarch quits at startup. E.g. 2560x240, unless you EDID-injected that modeline yourself. So making switchres work on Retroarch in DRM/KMS would imply getting into the DRM/KMS code of Retroarch rather than wrapping libswitchres.so? This also means that the user needs to EDID-inject CRT modelines.

That's almost it yes. I'm not sure RA parses the EDID, but KMS does it, yes. Adding user resolution to KMS has been removed a long time ago in kernel. Even if it could be "restored", it would mean having a specific kernel to get it working, and this is not the philosophy here.

Injecting an EDID means running RA as root or having some udev rules to change some permissions so this can be done at user level. I don't know well RA, so I'm not sure if EDID could be injected inside RA.

Anyway, the trick is more like being able to specify a modeline to RA's KMS code as RA can set a mode when creating a CRTC, this means a number of changes on SR side + specific code for KMS inside RA. But don't worry, I do want KMS modeswitching working 😉 This may not happen with that PR, but maybe rather in a near future. We have a bunch of errors to fix due to the variety of platforms RA cn run on. Then I'll look into KMS, I have a few leads to achieve that.

LarsenDX commented:
> > I looked into this some more. I understand the difficulty. Correct me if I am wrong. Retroarch in DRM/KMS is the DRM-Master. At startup, Retroarch queries for EDID modelines and tries to match video_fullscreen_x/video_fullscreen_y to a modeline. If it cannot find one, Retroarch quits at startup. E.g. 2560x240, unless you EDID-injected that modeline yourself. So making switchres work on Retroarch in DRM/KMS would imply getting into the DRM/KMS code of Retroarch rather than wrapping libswitchres.so? This also means that the user needs to EDID-inject CRT modelines.
> 
> That's almost it yes. I'm not sure RA parses the EDID, but KMS does it, yes. Adding user resolution to KMS has been removed a long time ago in kernel. Even if it could be "restored", it would mean having a specific kernel to get it working, and this is not the philosophy here.
> 
> Injecting an EDID means running RA as root or having some udev rules to change some permissions so this can be done at user level. I don't know well RA, so I'm not sure if EDID could be injected inside RA.
> 
> Anyway, the trick is more like being able to specify a modeline to RA's KMS code as RA can set a mode when creating a CRTC, this means a number of changes on SR side + specific code for KMS inside RA. But don't worry, I do want KMS modeswitching working 😉 This may not happen with that PR, but maybe rather in a near future. We have a bunch of errors to fix due to the variety of platforms RA cn run on. Then I'll look into KMS, I have a few leads to achieve that.

Indeed, that's the trick and that's what we should focus on. Happy to help, build, test or PR code if necessary. :)

alphanu1 commented:
Switchres is now baked in to RA.

Currently, DRM switching is not available. It wasn't before. However, this is in the works but could be weeks or months before its ready.

Currently, there is some videocore checks to disable switchres. The old RPi implementation is still in place. However, this will be ported into switchres. I will be working on this next. However, again it could be weeks or months before its ready.  

Last thing, switchres does have a ini. Many option are available within the RA UI. So the ini is not technically required. However, it does give the user more customisation options like monitor profiles, custom CRT ranges and options for fixing issues with specific video cards. I think ideally this should be a downloadable asset. Could the switchres.ini be supplied via the content downloader should the user require it? 

FYI, switchres and RA implementation is ongoing and a lot more UI option will be available down the road. However, as CRTSwithRes currently does not work correctly. This current release has mostly the same functionality with some extras. the only missing option is the x centering. However, this can be done using the switchres.ini for now if it is required.

@hunterk 
@twinaphex

hizzlekizzle commented:
>Could the switchres.ini be supplied via the content downloader should the user require it?

Sure, that's easy to do.

alphanu1 commented:
@hunterk  Currently, SR will look in the working directory or ./ini folder. This could be changed if needed. however, if the content downloader can download to one of these that would be great.

hizzlekizzle commented:
Unfortunately, the downloader is very limited and can only pull things to RetroArch's 'downloads' directory unless you set up a special case like the asset/overlay/shader updaters. Probably the easiest thing would be to just also look in 'downloads', though I don't know if you would want it to be higher or lower priority than the other locations (presumably lower, so people could download it and then override it with a custom one in the other spots).

alphanu1 commented:
As long as the user does not change the default location I guess.

Is there a different default location for windows? I know Linux is ~/.config/retroarch/downloads. As long as I can pull the downloads  location from RA it should be fine.

hizzlekizzle commented:
I'm not on a Windows machine at the moment, but i'm pretty sure it's just [RetroArch dir]\downloads there, too, but yeah, there's definitely a variable for it somewhere, since you can change the location in settings > directory.

alphanu1 commented:
Bug was found causing seg faults on RA exit. Calling RARCH_LOG to inform that SR will be deinit and then SR has been destroyed. in function `retroarch_deinit_drivers`

Seems that call RARCH logging in `retroarch_deinit_drivers` does not work and causes issues. Unable to inform user that SR has been deinit.

Logging SR in` retroarch_deinit_drivers` has now been disabled. However, This bug probably should be looked into.

alphanu1 commented:
Don't merge yet. Still a few bug reports coming in.

alphanu1 commented:
All the latest bug reports have now been fixed. You can go ahead and merge this when you are ready.

alphanu1 commented:
@hizzlekizzle I need to add a doc for this. Is it a simple case of adding a .md to the docs git? or is there a lot more involved?

hizzlekizzle commented:
That's pretty much it, yeah. Just write it up and PR it to the docs repo.

Kopert commented:
I'd like to move my test results and bug reports over here from #12244 as requested by @alphanu1. For context, I have been using SwitchRes on Linux for over a year now, and I have been following closely and testing builds for the repo that generated this PR for over a month now. My testings and opinions are based on comparisons to functionality from RetroArch 1.9.0, which is the last version where the old SwitchRes worked fine.

As of vesion 0.7, I consider this implementation to be a good replacement for the system we had before. There are lots of advantages from the development point of view, but I'll focus on the user side of things.

Pros:
- The whole resolution switching process seems to be a bit faster and to behave more predictably
- Since its based on an external implementation, it is independently documented and supported, which is helpful for people needing assistance with specific issues
- Being able to edit the INI file to adjust some settings and, more importantly, set custom display profiles, allows the user to do things that weren't possible before, like running content that has a native refresh rate higher than the display in use or customizing porch settings to fix horizontal overscan in displays that don't have that sort of control

Cons:
- Some stuff might not work exactly as before if you don't adjust settings on the INI file
- The documentation for any of this is not part of the RA documentation yet
- There are a couple of minor bugs

I believe the pros overweight the cons, specially considering that simply merging this PR would get most users back to having a working SwitchRes (which hasn't been the case since 1.9.1). The INI file cons are easily fixable with documentation and, in the future, with adding more options to the RA UI.

The two minor bugs I can still find in the code (that were not present in the 1.9.0 implementation) are:
- Handheld content does not have the proper aspect ratio set. 1.9.0 was inconsistent with this and most handheld cores needed some adjustment anyway, so users who care about this (like me) probably already had made some adjustments themselves (which, in my case, still work on the new implementation). I've been told that a proper implementation for this is being done in the SwitchRes library itself.
- When using super resolutions of 1920 and 2560, 288p50hz content is being shown in 576i50hz instead of the original 288p. This can be worked around by using a super resolution of 3840, or NATIVE/DYNAMIC.

I tested multiple types of content loaded in a couple of different ways and could not find any other issues on 0.7.

My recommendation at this point with version 0.7 would be to move forward with the merge so other users can go back to having SwitchRes working, as the bugs are minor and easy to work around. The first documentation efforts should talk about how to use the switchres.ini file and how to use monitor presets with it, as this should help to iron out most initial issues that might crop up.

twinaphex commented:
Hi @Kopert, can you help by updating the documentation on the docs repo?

Kopert commented:
> 
> 
> Hi @Kopert, can you help by updating the documentation on the docs repo?

@twinaphex I don't have a lot of experience with this stuff but I'd be glad to try. Should I just create a file called `docs/guides/switchres.md` on a fork to PR later? Is @alphanu1 ok with me doing this?

alphanu1 commented:
@twinaphex I will be creating a docs page over on the docs repo.  Once I done this I'll get @Kopert to have a look though it and make sure I have not missed anything. 

I have an update for PI CRT switching to which I will.push later today. 

So give me a couple.of days and everything will be ready., with docs.

alphanu1 commented:
@hunterk @twinaphex Current guide for [CRTSwithRes](https://github.com/alphanu1/MME4CRT_docs/blob/master/docs/guides/crtswitchres.md)

alphanu1 commented:
Fixed file conflicts

alphanu1 commented:
Few more bug found and fixed 😄

Sukid commented:
Will this be usable with wayland compositors as well? I'm not sure if Wayland on its own has an interface for doing this but I know Sway can do it, and it's based on wlroots. I'm pretty sure any wayland application using wlroots would probably work fine even in non-wlroots based compositors like KDE and GNOME if gamescope is any indication.

alphanu1 commented:
We have not currently been looking at Wayland. However, that's not to say it will never happen. 

There will be a PR shortly after this is merged from @substring with something that we though would probably not happen, props to him.

alphanu1 commented:
@hunterk @twinaphex 

Rebased and fixed the new file conflicts. Just waiting on this merge so we can implement KMS/DRM mode switching.

FYI, We do already have KMS/DRM mode switching working. Making it compatible with any console Linux.

FroyoXSG commented:
Hi RetroArch team and contributors, if you're already updating that part of the documentation, consider clarifying whether Windows users *need* to have CRTEmudriver. I've had success with 2560 super res on unmodified Nvidia Drivers. Works about 90%, with the occasional failure where switching to interlaced fails/wasn't done and it displays a progressive image with very large pillarboxes.

DisasterMo commented on intl/msg_hash_us.h:
```diff
@@ -1437,9 +1437,17 @@ MSG_HASH(
    MENU_ENUM_SUBLABEL_CRT_SWITCH_PORCH_ADJUST,
    "Cycle through these options to adjust the porch settings to change the image size."
    )
+MSG_HASH(
+   MENU_ENUM_LABEL_VALUE_CRT_SWITCH_HIRES_MENU,
+   "Use high resolution menu"
+   )
+MSG_HASH(
+   MENU_ENUM_SUBLABEL_CRT_SWITCH_HIRES_MENU,
+   "Use high resolution menu"
+   )
```
Please don't create redundant sublabels. Please try to describe the function and/or purpose and/or when/how to use this function in more detail. Either make the sublabel be useful, in any way, or don't add one. Thank you.

DisasterMo commented on intl/msg_hash_us.h:
```diff
@@ -1437,9 +1437,17 @@ MSG_HASH(
+   "Use high resolution menu"
+   )
 MSG_HASH(
    MENU_ENUM_LABEL_VALUE_CRT_SWITCH_RESOLUTION_USE_CUSTOM_REFRESH_RATE,
-   "Use Custom Refresh Rate"
+   "Use a custom refresh rate specified in the configuration file if needed."
    )
 MSG_HASH(
    MENU_ENUM_SUBLABEL_CRT_SWITCH_RESOLUTION_USE_CUSTOM_REFRESH_RATE,
```
Why? Why would you replace a label with it's sublabel? And introdunce redundancy by not changing the sublabel? What is even going on here?

alphanu1 commented:
@DisasterMo These have both been sorted. and will be PRd in the next update.

Not sure what happened with Custom Refresh Rate. but not to worry its all sorted.

@FroyoXSG The docs should be updated soon.

DisasterMo commented:
Okay then, good work!
Please excuse my 'tone' (I guess?), it's just that every change to the msg_hash_us.h means extra work for the translators, so any unneccessary changes should be avoided as much as possible. Thank you!

sonninnos commented:
Unfortunately that addition in winraw_input.c produces unlimited amount of 
```
Exception code=0xc0000005 flags=0x0 at 0x00007FFC3A5C2EFB. Access violation - attempting to write data at address 0x0000000000000000
```
on mouse move, and naturally it results in random behavior, such as erratic menu mouse.

alexb3d commented:
Tested in the new version 1.9.7.
It works, although on a monitor presents some errors. I have not been able to use it on a TV.

1. You have to use the advanced menu, activate the option in Basic Menu does not do anything.

2. You have to activate the scaling video option, because if not, [the image is crushed.](https://ibb.co/WDWywd5)

3. When closed, return to the resolution of the desktop correctly. But when it restarts and then closes, it stays at 240

4. When the game is loaded does not activate 240P, you have to close the game and reload it.

5. Activating the menu works well without loading a game, when the game and the menu is activated, [you can not distinguish.](https://ibb.co/7yLCyJR)
